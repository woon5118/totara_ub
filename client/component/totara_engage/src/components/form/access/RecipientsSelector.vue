<!--
  This file is part of Totara Enterprise Extensions.

  Copyright (C) 2020 onwards Totara Learning Solutions LTD

  Totara Enterprise Extensions is provided only to Totara
  Learning Solutions LTD's customers and partners, pursuant to
  the terms and conditions of a separate agreement with Totara
  Learning Solutions LTD or its affiliate.

  If you do not have an agreement with Totara Learning Solutions
  LTD, you may not access, use, modify, or distribute this software.
  Please contact [licensing@totaralearning.com] for more information.

  @author Johannes Cilliers <johannes.cilliers@totaralearning.com>
  @module totara_engage
-->

<template>
  <div class="tui-engageSharedRecipientsSelector">
    <Label
      :label="shareLabel"
      :for-id="generatedId"
      class="tui-engageSharedRecipientsSelector__label"
    />
    <InfoIconButton
      :is-help-for="shareLabel"
      class="tui-engageSharedRecipientsSelector__icon"
    >
      {{ shareHelpInfo }}
    </InfoIconButton>
    <TagList
      :id="generatedId"
      :filter="query"
      :items="items"
      :tags="tags"
      :separator="true"
      @filter="find"
      @select="select"
      @remove="remove"
    >
      <template v-slot:item="{ item }">
        <div
          v-if="item.user"
          class="tui-engageSharedRecipientsSelector__profileContainer"
        >
          <MiniProfileCard
            :no-border="true"
            :no-padding="true"
            :read-only="true"
            :display="item.user.card_display"
          />
          <div
            v-if="getRecipientDetails(item, 'alreadyshared')"
            class="tui-engageSharedRecipientsSelector__profileContainer-badge"
          >
            <CheckSuccess />
            <span>{{ $str('alreadyshared', 'totara_engage') }}</span>
          </div>
        </div>

        <div v-else class="tui-engageSharedRecipientsSelector__recipient">
          <Avatar
            :src="getRecipientDetails(item, 'src')"
            :alt="getRecipientDetails(item, 'fullname')"
            size="xsmall"
          />
          <ul class="tui-engageSharedRecipientsSelector__recipient-summary">
            <li>{{ getRecipientDetails(item, 'fullname') }}</li>
            <li>{{ getRecipientDetails(item, 'summary') }}</li>
          </ul>
          <div
            v-if="getRecipientDetails(item, 'alreadyshared')"
            class="tui-engageSharedRecipientsSelector__recipient-badge"
          >
            <CheckSuccess />
            <span>{{ $str('alreadyshared', 'totara_engage') }}</span>
          </div>
        </div>
      </template>
    </TagList>
  </div>
</template>

<script>
import MiniProfileCard from 'tui/components/profile/MiniProfileCard';
import Avatar from 'tui/components/avatar/Avatar';
import InfoIconButton from 'tui/components/buttons/InfoIconButton';
import CheckSuccess from 'tui/components/icons/CheckSuccess';
import Label from 'tui/components/form/Label';
import TagList from 'tui/components/tag/TagList';
import { AccessManager } from 'totara_engage/index';
import { config } from 'tui/config';

// GraphQL
import shareToRecipients from 'totara_engage/graphql/shareto_recipients';
import engageAdvancedFeatures from 'totara_engage/graphql/advanced_features';

// Mixin
import ContainerMixin from 'totara_engage/mixins/container_mixin';

export default {
  components: {
    Avatar,
    InfoIconButton,
    Label,
    TagList,
    CheckSuccess,
    MiniProfileCard,
  },

  mixins: [ContainerMixin],

  props: {
    itemId: {
      type: [Number, String],
      required: true,
    },

    component: {
      type: String,
      required: true,
    },

    access: {
      type: String,
      default: null,
      validator(prop) {
        return AccessManager.isValid(prop);
      },
    },

    owned: {
      type: Boolean,
      default: true,
    },
  },

  apollo: {
    features: {
      query: engageAdvancedFeatures,
    },

    recipients: {
      query: shareToRecipients,
      fetchPolicy: 'network-only',
      variables() {
        return {
          itemid: this.itemId,
          component: this.component,
          search: this.query,
          access: this.access,
          theme: config.theme.name,
        };
      },

      result({ data: { recipients } }) {
        let tmp_recipients = [];

        // As preventing a tag from being manually removed isn't currently possible
        // we do not filter the this.containerRecipient item out.

        // ID is not enough to uniquely identify a specific recipient.
        tmp_recipients = recipients.map(recipient => {
          return Object.assign({}, recipient, {
            id:
              recipient.component +
              '/' +
              recipient.area +
              '/' +
              recipient.instanceid,
          });
        });

        // Filter the null value users
        tmp_recipients = tmp_recipients.filter(recipient => {
          if (recipient.area === 'user') {
            return recipient.user.card_display.display_fields.some(
              field => field.value != null
            );
          }
          // Return all workspaces as true
          return true;
        });

        this.recipients = tmp_recipients;
      },
    },
  },

  data() {
    return {
      query: '',
      recipients: [],
      publicTags: [],
      restrictedTags: [],
      features: {},
    };
  },

  computed: {
    tags() {
      return this.accessPublic ? this.publicTags : this.restrictedTags;
    },

    items() {
      return this.recipients.filter(
        recipient => !this.tags.some(tag => recipient.id === tag.id)
      );
    },

    containerRecipient() {
      if (this.container && this.container.autoShareRecipient) {
        return {
          instanceid: this.containerValues.instanceId,
          component: this.containerValues.component,
          area: this.containerValues.area,
          minimum_access: this.containerValues.access,
          other: {
            fullname: this.containerValues.name,
          },
        };
      }
      return false;
    },

    generatedId() {
      return this.$id();
    },

    featureWorkspaces() {
      return this.features && this.features.workspaces;
    },

    shareLabel() {
      const labels = this.featureWorkspaces
        ? [
            'resharetorecipientsworkspaces',
            'sharetorecipientsworkspaces',
            'sharetorecipientsoptionalworkspaces',
          ]
        : [
            'resharetorecipients',
            'sharetorecipients',
            'sharetorecipientsoptional',
          ];

      if (!this.owned) {
        return this.$str(labels[0], 'totara_engage');
      } else {
        return this.$str(
          this.accessRestricted ? labels[1] : labels[2],
          'totara_engage'
        );
      }
    },

    shareHelpInfo() {
      return this.$str(
        this.featureWorkspaces ? 'sharehelpinfoworkspaces' : 'sharehelpinfo',
        'totara_engage'
      );
    },

    accessPublic() {
      return AccessManager.isPublic(this.access);
    },

    accessRestricted() {
      return AccessManager.isRestricted(this.access);
    },
  },

  mounted() {
    if (this.containerRecipient) {
      this.select(this.containerRecipient);
    }
  },

  methods: {
    /**
     * @param {String} query
     */
    find(query) {
      if ('' === query) {
        this.query = query;
        return;
      }

      this.query = query;
      this.skip = false;

      this.$apollo.queries.recipients.refetch();
    },

    /**
     *
     * @param {Object} item
     */
    select(item) {
      if (item.alreadyshared) {
        // ignore already shared users
        return;
      }

      const tag = this.createTag(item);
      this.publicTags.push(tag);
      if (AccessManager.isRestricted(item.minimum_access)) {
        this.restrictedTags.push(tag);
      }

      this.$emit('pick-recipient', item);
      // Reset query to empty string
      this.query = '';
    },

    /**
     *
     * @param {Object} tag
     */
    remove(tag) {
      this.publicTags = this.publicTags.filter(t => t !== tag);
      if (AccessManager.isRestricted(tag.minimum_access)) {
        this.restrictedTags = this.restrictedTags.filter(t => t !== tag);
      }
      this.$emit('remove-recipient', tag);
    },

    /**
     *
     * @param recipient
     * @param type
     * @returns {*}
     */
    getRecipientDetails(recipient, type) {
      // Shared properties.
      switch (type) {
        case 'alreadyshared':
          return recipient.alreadyshared;
        case 'summary':
          return recipient.summary;
      }

      // User specific. equal to recipient.area === 'user' && type === 'fullname'
      if (recipient.area === 'user') {
        const displayField =
          recipient.user.card_display.display_fields.find(
            field => field.label === 'Full name'
          ) || recipient.user.card_display.display_fields[0];
        return displayField.value;
      } else {
        // Currently it's for workspace
        switch (type) {
          case 'src':
            return recipient.other.imageurl || '';
          case 'alt':
            return recipient.other.imagealt || '';
          case 'fullname':
            return recipient.other.fullname;
        }
      }
    },

    /**
     * @param {Object} recipient
     */
    createTag(recipient) {
      return {
        text: this.getRecipientDetails(recipient, 'fullname'),
        id:
          recipient.component +
          '/' +
          recipient.area +
          '/' +
          recipient.instanceid,
        instanceid: recipient.instanceid,
        area: recipient.area,
        component: recipient.component,
        minimum_access: recipient.minimum_access,
      };
    },
  },
};
</script>

<lang-strings>
  {
    "totara_engage": [
      "alreadyshared",
      "resharetorecipients",
      "resharetorecipientsworkspaces",
      "sharehelpinfo",
      "sharehelpinfoworkspaces",
      "sharetorecipients",
      "sharetorecipientsworkspaces",
      "sharetorecipientsoptional",
      "sharetorecipientsoptionalworkspaces"
    ]
  }
</lang-strings>

<style lang="scss">
.tui-engageSharedRecipientsSelector {
  &__label.tui-formLabel {
    @include tui-font-heading-label();
    margin-right: 0;
    margin-bottom: var(--gap-2);
  }

  &__icon {
    display: inline-flex;
  }

  &__recipient {
    display: flex;
    > :first-child {
      margin-right: var(--gap-2);
    }

    &-summary {
      margin: 0;
      list-style-type: none;
      > :first-child {
        @include tui-font-heading-label();
      }
      > :last-child {
        @include tui-font-body-small();
      }
    }

    &-badge {
      align-self: flex-end;
      margin-left: auto;
      > :last-child {
        @include tui-font-body-small();
      }
    }
  }

  &__profileContainer {
    position: relative;
    &-badge {
      display: flex;
      align-items: flex-end;
      justify-content: flex-end;

      > :last-child {
        @include tui-font-body-small();

        padding-left: var(--gap-1);
      }
    }
  }
}
</style>
