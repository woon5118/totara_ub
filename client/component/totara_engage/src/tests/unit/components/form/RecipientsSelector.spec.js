/**
 * This file is part of Totara Enterprise Extensions.
 *
 * Copyright (C) 2020 onwards Totara Learning Solutions LTD
 *
 * Totara Enterprise Extensions is provided only to Totara
 * Learning Solutions LTD's customers and partners, pursuant to
 * the terms and conditions of a separate agreement with Totara
 * Learning Solutions LTD or its affiliate.
 *
 * If you do not have an agreement with Totara Learning Solutions
 * LTD, you may not access, use, modify, or distribute this software.
 * Please contact [licensing@totaralearning.com] for more information.
 *
 * @author Brian Barnes <brian.barnes@totaralearning.com>
 * @module totara_engage
 */

import { mount } from '@vue/test-utils';
import RecipientsSelector from 'totara_engage/components/form/access/RecipientsSelector';

describe('totara_engage/components/form/RecipientsSelector.vue', () => {
  let wrapper = null;

  beforeAll(() => {
    wrapper = mount(RecipientsSelector, {
      mocks: {
        $str(id, component) {
          return `${id}, ${component}`;
        },
        $id() {
          return 'abc123';
        },
      },
      propsData: {
        itemId: 0,
        component: 'engage_resource',
        access: 'PUBLIC',
        selectedRecipients: [],
        owned: false,
      },
    });
  });

  it('matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('selecting a user works as expected', () => {
    let item = {
      area: 'user',
      alreadyshared: false,
      user: {
        card_display: {
          display_fields: [
            {
              label: 'Full name',
              value: 'User 1',
            },
          ],
        },
      },
    };
    wrapper.vm.select(item);
    expect(wrapper.emitted()['pick-recipient'].length).toBe(1);

    // If a user is already selected then they should not be selected again
    let item2 = {
      area: 'user',
      alreadyshared: true,
      user: {
        card_display: {
          display_fields: [
            {
              label: 'Full name',
              value: 'User 2',
            },
          ],
        },
      },
    };
    wrapper.vm.select(item2);
    expect(wrapper.emitted()['pick-recipient'].length).toBe(1);

    // Test other recipient.
    let item3 = {
      area: 'other',
      alreadyshared: false,
      other: {
        fullname: 'Other 1',
      },
    };
    wrapper.vm.select(item3);
    expect(wrapper.emitted()['pick-recipient'].length).toBe(2);
  });
});
