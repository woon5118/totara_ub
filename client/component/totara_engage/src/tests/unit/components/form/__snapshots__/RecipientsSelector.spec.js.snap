// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`totara_engage/components/form/RecipientsSelector.vue matches snapshot 1`] = `
<div class="tui-engageSharedRecipientsSelector" selectedrecipients=""><label for="abc123" class="tui-formLabel tui-engageSharedRecipientsSelector__label">
    <!---->

    resharetorecipients, totara_engage
    <!---->
  </label>
  <div class="tui-engageSharedRecipientsSelector__icon"><button aria-expanded="false" aria-label="show_help_for_x, totara_core" type="button" title="show_help_for_x, totara_core" class="tui-iconBtn tui-infoIconButton tui-iconBtn--prim tui-iconBtn--small tui-iconBtn--transparent"><span class="tui-iconBtn__wrap"><span class="tui-iconBtn__label"><span aria-hidden="true" class="tui-iconBtn__icon"><span class=""></span></span>
      <!----></span>
      <!---->
      <!----></span>
    </button>
    <div class="tui-popoverPositioner tui-popoverPositioner--transition-default" style="transform: translate3d(0px, 0px, 0); display: none;">
      <div class="tui-popoverFrame tui-popoverFrame--bottom">
        <div class="tui-arrow tui-arrow--top tui-arrow--normal" style="left: 0px;"></div>
        <!---->
        <!---->
        <div role="tooltip" class="tui-popoverFrame__content tui-popoverFrame__content--nonClosable">
          sharehelpinfo, totara_engage
        </div>
        <!---->
      </div>
    </div>
  </div>
  <div class="tui-dropdown tui-dropdown--flexibleWidth" id="abc123">
    <div id="abc123" class="tui-dropdown__trigger">
      <div class="tui-tagList">
        <div class="tui-tagList__tags">
          <div aria-live="polite" role="status" aria-atomic="false" aria-label="tags_selected, totara_core" aria-relevant="additions" class="tui-tagList__tagItems">
            <div class="tui-tagList__input"><input aria-label="tag_list, totara_core" placeholder="tag_list_placeholder, totara_core" type="text" class="tui-formInput tui-formInput--transparent"></div>
          </div> <span class="tui-tagList__suffix" style="display: none;">n_more, totara_core</span>
        </div> <button aria-expanded="false" aria-haspopup="menu" aria-label="expand, core tag_list, totara_core" type="button" title="expand, core tag_list, totara_core" class="tui-iconBtn tui-tagList__expandArrow tui-iconBtn--transparent"><span class="tui-iconBtn__wrap"><span class="tui-iconBtn__label"><span aria-hidden="true" class="tui-iconBtn__icon"><span class=""></span></span>
          <!----></span>
          <!---->
          <!----></span>
        </button>
      </div>
    </div>
    <!---->
  </div>
</div>
`;
